packages/doenetml-worker-rust/lib-doenetml-core/src/graph/directed_graph/graph_walk.rs

```diff
@@ -0,0 +1,160 @@
+
+        DescendantTopologicalIterator::new_multiroot(&self.nodes, &self.edges, start_indices)
+    }
+
+    /// Walk through all nodes that have any node listed in `start_nodes` as an ancestor, including `start_nodes` themselves.
+    /// Nodes are walked in _reverse topological_ order.
+    /// That is, if there is an edge `a -> b`, the node `b` will be visited before `a`.
+    /// Panics if a cycle is detected.
+    pub fn descendants_reverse_topological_multiroot<A: Borrow<Node>>(
+        &self,
+        start_nodes: &[A],
+    ) -> impl Iterator<Item = &Node> {
+        let start_indices = start_nodes
+            .iter()
+            .map(|node| *self.index_lookup.get_tag(node.borrow()).unwrap())
+            .collect();
+        DescendantReverseTopologicalIterator::new_multiroot(&self.nodes, &self.edges, start_indices)
+    }
+
+    /// Walk through all nodes that have any node listed in `start_nodes` as an ancestor, including `start_nodes` themselves.
+    /// Nodes for which `skip` returns `true` are treated as if they are absent from the graph.
+    /// Nodes are walked in _reverse topological_ order.
+    /// That is, if there is an edge `a -> b`, the node `b` will be visited before `a`.
+    /// Panics if a cycle is detected.
+    pub fn descendants_reverse_topological_multiroot_with_skip<
+        A: Borrow<Node>,
+        SkipFn: Fn(&Node) -> bool,
+    >(
+        &self,
+        start_nodes: &[A],
+        skip: SkipFn,
+    ) -> impl Iterator<Item = &Node> {
+        let start_indices = start_nodes
+            .iter()
+            .map(|node| *self.index_lookup.get_tag(node.borrow()).unwrap())
+            .collect();
+        DescendantReverseTopologicalIterator::new_multiroot_with_skip(
+            &self.nodes,
+            &self.edges,
+            start_indices,
+            skip,
+        )
+    }
+
+    /// Walk through all nodes that have any node listed in `start_nodes` as a descendant, including `start_nodes` themselves.
+    /// Nodes are walked in _topological_ order.
+    /// That is, if there is an edge `a -> b`, the node `a` will be visited before `b`.
+    /// Panics if a cycle is detected.
+    pub fn ancestors_topological_multiroot<A: Borrow<Node>>(
+        &self,
+        start_nodes: &[A],
+    ) -> DescendantTopologicalIterator<Node> {
+        let start_indices = start_nodes
+            .iter()
+            .map(|node| *self.index_lookup.get_tag(node.borrow()).unwrap())
+            .collect();
+
+        DescendantTopologicalIterator::new_multiroot(
+            &self.nodes,
+            &self.reverse_edges,
+            start_indices,
+        )
+    }
+
+    /// Walk through all nodes that have `node` as a descendant, including `node` itself.
+    /// Nodes are walked in _reverse topological_ order.
+    /// That is, if there is an edge `a -> b`, the node `b` will be visited before `a`.
+    /// Panics if a cycle is detected.
+    pub fn ancestors_reverse_topological<A: Borrow<Node>>(
```

Comment:
Did you mean to keep this one that isn't multiroot?

Reply:
Nope. Just forgot it!